### ISSUE TITLE
inject optional parameters

### ISSUE THREAD

**Issue by Harshita-mindfire** [microsoft/tsyringe#207]
Extremely new to the DI tsyringe, how do we handle optional parameters with`@inject` decorator?
With the below setup, I get the following error:
```
Cannot inject the dependency siteUtils in UtilsWeb, Reason: TypeInfo not known for "undefined"
```
My use case is : the siteUrl and enablePrettyLinks can have undefined values. They are being read from a .yml file
```ts
-------------
  container.register("siteUrl", { useValue: siteUrl }); // siteUrl can have undefined value
  container.register("enablePrettyLinks", { useValue: enablePrettyLinks }); // enablePrettyLinks can have undefined value
------------------

@injectable()
export class UtilsWeb {
  constructor(
    @inject("siteUrl") private siteUrl?: string,
    @inject("enablePrettyLinks") private enablePrettyLinks?: boolean
  ) {}
...
}
```

**Comment by etiennenoel** [comment]
Sorry that we are just starting to look at this. What if we added an optional parameter to `@inject("", true)` to follow this: https://github.com/microsoft/tsyringe/pull/252

Thinking about it, maybe a better idea is an optional object, something like:

```
@inject("", { isOptional?: boolean });
@injectAll("", { isOptional?: boolean });
```

@Xapphire13 Any thoughts?

**Comment by etiennenoel** [comment]
or maybe we provide even more flexibility and go with a `defaultValue` approach:

```
@inject("", { defaultValue?: any });
@injectAll("", { defaultValue?: any });
```

If defaultValue is not specified, this would mean that it's not optional. Is that behaviour sufficiently clear enough for developers to understand what it does?

**Comment by etiennenoel** [comment]
I agree. I also think I should revise my original PR and use an option object as a second argument. That way, it might be easier to add future parameters and/or deprecate some while being retro-compatible.

**Comment by etiennenoel** [comment]
Will be available in next release!
